### PR TITLE
Fix MatMul AutoSchedule Precision Problem

### PR DESCRIPTION
**Pull request by zhhsplendid**
This PR fixed 3 problems which can cause AutoSchedule precision problem.

3 Problems:
1. `__launch_bounds__` is set to 1 wrongly sometimes. The reason of bug is that `__launch_bounds__` information is added in `optim::Optimize` (I know, it is not a good name that doing fundamental data setting in a function calling Optimize, we will fix it later), our solution is moving `PrepareCudaAxisInfoFromBody` of LoweredFunc before `optim::Optimize`
2. `threadIdx` can exceed the limit of CUDA, we set 1024 as the upper bound limit.
3. Matrix reduce init statements lost during Reorder schedule primitive, we fixed it and changed related MulitLevelTiling code. 

We manually tested the log CUDA kernel that precision is accurate after merging this PR.

### CODE REVIEW

File: cinn/auto_schedule/tests/performance_comparison_test.cc [hunk 1]
```diff
@@ -186,6 +186,7 @@ TEST_F(PerformanceTester, Matmul) {
   int K = 2048;
   int N = 1000;
 
+  options_.evaluate_knobs = 4UL;
```
Comment (CtfGo) [inline cinn/auto_schedule/tests/performance_comparison_test.cc]: 为什么要限制为4

File: cinn/ir/ir_schedule.cc [hunk 2]
```diff
@@ -855,8 +855,8 @@ void ScheduleImpl::SyncThreads(const Expr& ir_node, bool after_node) {
  * @param tgt_stmt The For node we want.
  */
 void ScheduleImpl::Replace(const Expr& src_sref, const Expr& tgt_stmt) {
-  CHECK((src_sref.As<ir::For>() && tgt_stmt.As<ir::For>()) || (src_sref.As<ir::Block>() && tgt_stmt.As<ir::Block>()) ||
-        (src_sref.As<ir::ScheduleBlockRealize>() && tgt_stmt.As<ir::ScheduleBlockRealize>()));
+  CHECK(src_sref.As<ir::For>() || src_sref.As<ir::Block>() || src_sref.As<ir::ScheduleBlockRealize>());
+  CHECK(tgt_stmt.As<ir::For>() || tgt_stmt.As<ir::Block>() || tgt_stmt.As<ir::ScheduleBlockRealize>());
```
Comment (CtfGo) [inline cinn/ir/ir_schedule.cc:859]: 这个修改是什么原因呢？允许两个不同类型的Expr替换吗

File: cinn/ir/ir_schedule.h [hunk 3]
```diff
@@ -201,22 +201,25 @@ class IRSchedule {
   /**
    * \brief Reorder the loops in the order of vector.
    * @param loops The loops to be reordered.
+   * @return The reordered Expr, can be ir::For or ir::Block.
```
Comment (CtfGo) [inline cinn/ir/ir_schedule.h]: 可能返回两种类型，For和Block分别对应什么结果？什么情况下会返回for/block
能否依旧返回void，使用者再通过其它接口get对应的for或block

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc [hunk 4]
```diff
@@ -186,16 +186,81 @@ void MultiLevelTiling::Apply(int index) {
       loop_var_name_to_idx[for_exprs[j].As<ir::For>()->loop_var->name] = j;
     }
     CHECK(loop_var_name_to_idx.size() == for_exprs.size()) << "Loops contain duplicate loop var names before Fusion";
+
+    int extent_prod                    = 1;
+    int first_idx_less_than_max_factor = -1;
     for (int j = 0; j < tiles[i].size(); ++j) {
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc:161]: 关键步骤代码添加说明注释

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc [hunk 5]
```diff
@@ -186,16 +186,81 @@ void MultiLevelTiling::Apply(int index) {
...
+    }
+  }
+
+  VLOG(5) << "Do Fuse and Binding on the non-main loop chains";
+  Expr sche_block_top_loop = ir_schedule_->GetLoops(sche_block->name)[0];
+
+  std::vector<Expr> scan_loop_blocks = ir_schedule_->GetAllBlocks();
+  for (const Expr& b : scan_loop_blocks) {
+    ir_schedule_->GetLoops(b);
+  }
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc]: 这几句是什么作用

Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc]: 起初是因为现在对schedule的操作必须是其他schedule的输出，所以在初次把block和loop都拿出来。再次review发现可以把这几句删掉，是冗余代码，因为下面有做同样的拿block和拿loops

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc (hunk 4 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc:161]: Done

File: cinn/ir/ir_schedule.h (hunk 3 above)
Comment (zhhsplendid) [reply cinn/ir/ir_schedule.h]: Done

File: cinn/ir/ir_schedule.cc (hunk 2 above)
Comment (zhhsplendid) [reply cinn/ir/ir_schedule.cc:859]: 是的，需要允许，否则现在比如Reorder后含有多条loop chain，这种情况无法处理。

File: cinn/auto_schedule/tests/performance_comparison_test.cc (hunk 1 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/tests/performance_comparison_test.cc]: 自己测试时代码没删干净。现在把这句删了

**Review by CtfGo — APPROVED** [review]
LGTM

**Pull request merged**